### PR TITLE
add machine controller first iteration

### PR DESCRIPTION
**Pull request by iamkirkbater**
Template from @maorfr in #112 

[issues.redhat.com/browse/OSD-18411](https://issues.redhat.com/browse/OSD-18411)

Goal here is to detect customer pods failing to drain from deleting machines in order to send SLs through OCM Agent. This would remove SRE action from any customer-pod-based KubeNodeUnscheduleable alert.

Example metrics 
```
# HELP pods_preventing_node_drain Pods that cannot be drained from a deleting machine
# TYPE pods_preventing_node_drain gauge
pods_preventing_node_drain{instance="ip-10-0-226-104.ec2.internal",machine="kirk-test-ccs-ltfm9-worker-us-east-1a-qrdmb",node="ip-10-0-226-104.ec2.internal",pod_name="gin-app-68d7f74fbd-cjxnx",pod_namespace="gin2"} 1
pods_preventing_node_drain{instance="ip-10-0-226-104.ec2.internal",machine="kirk-test-ccs-ltfm9-worker-us-east-1a-qrdmb",node="ip-10-0-226-104.ec2.internal",pod_name="postgresql-1-q5d7w",pod_namespace="ruby"} 1
```

The idea being that after this is rolled out and tested that KubeNodeUnscheduleableSRE would be modified to something like:

```
kubeNodeUnscheduleable[instance] for > 60m AND NOT count(pods_preventing_node_drain[instance]) > 0
```

The SL would be sent automatically with something like:

```
count(pods_preventing_node_drain) > 0 for 20m
```

Since there's no ability to template SLs at this point in time, we can point the customers to look at the prometheus metrics for their cluster in the ServiceLog, querying `pods_preventing_node_drain` to see all pods preventing deleting machines from draining across their cluster, and take action for those.

### CODE REVIEW

File: controllers/machine/machine_controller.go
Comment (mjlshen) [inline controllers/machine/machine_controller.go:1]: Will need to add additional K8s RBAC in `deploy/` as well

**Review by mjlshen — CHANGES_REQUESTED** [review]

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go:89]: Yes, this index is required in order for us to use the FIeldSelector on the `involvedObject.name` in order to "server-side" filter the events to the list of those only given off for the specific machine we're reconciling against.

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go:132]: Sure - that was my first thought as well - I just couldn't actually find any prior art on how to do that within go, granted I didn't look _too_ hard after my first initial searches came up empty.  

Because of that failure to find it, I actually was able to use a Test-Driven Development approach with this function, and was able to define the tests and write the function at the same time which was actually a nice little experience because I haven't done that in a long time haha :)

File: go.mod
Comment (iamkirkbater) [reply go.mod:8]: Hmm, I'm not sure specifically why this updated - I'll see if I can try to roll this change back.

File: pkg/metrics/aggregator.go
Comment (iamkirkbater) [reply pkg/metrics/aggregator.go:229]: Yep, you're right here. Looks like this was leftover from an earlier initialization process I had. I'll clean this up, good catch.

File: controllers/machine/machine_controller.go
Comment (AlexVulaj) [reply controllers/machine/machine_controller.go:132]: This is pseudocode written directly in this comment and not tested, but you should be able to do something like...

``` go
sort.Slice(eventList.Items, func(i, j event) bool {
  return eventList.Items[i].LastTimeStamp < eventList.Items[j].LastTimeStamp
})
```

File: pkg/metrics/aggregator.go
Comment (iamkirkbater) [inline pkg/metrics/aggregator.go:256]: TODO - Do we need to Reset and re-publish existing metrics here again?

File: controllers/machine/machine_controller.go
Comment (mjlshen) [reply controllers/machine/machine_controller.go:108]: Personally I think passing the machine explicitly is clearer as it avoids implicit dependencies (calling a function by passing a context without a machine in it when it's needed), so if there isn't a reason to smuggle it through `ctx` I'd prefer to be explicit - as you called out, not too many parameters yet!

File: controllers/machine/machine_controller.go
Comment (mjlshen) [reply controllers/machine/machine_controller.go:89]: Thanks, that makes sense

File: go.mod
Comment (iamkirkbater) [reply go.mod:8]: So after testing this today, this _is_ the 4.11 release.

```> go get github.com/openshift/api@release-4.11


> git diff


> go get github.com/openshift/api@release-4.12
go: downloading github.com/openshift/api v0.0.0-20230928134114-673ed0cfc7f1
go: upgraded github.com/openshift/api v0.0.0-20230426102702-398424d53f74 => v0.0.0-20230928134114-673ed0cfc7f1


> go get github.com/openshift/api@release-4.11
go: downgraded github.com/openshift/api v0.0.0-20230928134114-673ed0cfc7f1 => v0.0.0-20230426102702-398424d53f74

```

**Review by maorfr — APPROVED** [review]
:heart:

**Comment by maorfr** [comment]
/hold

i didn't intend to let this merge, just to show my support

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go:103]: I'll leave this up to y'all to tell me if this looks any simpler - I'm not seeing it as any easier to read but I'm also biased because I wrote it XD

```
	sort.Slice(eventList.Items, func(i, j int) bool {
		return eventList.Items[i].LastTimestamp.Time.After(eventList.Items[j].LastTimestamp.Time)
	})

	for i := range eventList.Items {
		event := &eventList.Items[i]
		if event.Reason == "DrainRequeued" {
			if strings.Contains(event.Message, "error when evicting pods") {
				newestEvent = event
				break
			}
		}
	}
```

File: controllers/machine/machine_controller.go
Comment (mrWinston) [reply controllers/machine/machine_controller.go:88]: @iamkirkbater do you mind adding this as a comment to the code? I was just asking myself the very same question :)

File: controllers/machine/machine_controller.go
Comment (mrWinston) [inline controllers/machine/machine_controller.go:159]: Can you make the 15 minutes a Constant so that it's easier to adjust in the future?

File: controllers/machine/machine_controller.go
Comment (mrWinston) [inline controllers/machine/machine_controller.go:174]: I'm wondering if we should check weather the `event.LastTimestamp > machine.DeletionTimestamp`...
Is it even possible that we find a drain failed event for a previous deletion attempt?

File: controllers/machine/machine_controller.go
Comment (mrWinston) [reply controllers/machine/machine_controller.go:103]: I'd say leave the loop as is. I personally find this approach a bit easier to understand. But now that we're on the topic of readability:
I would like to see the conditions of the if-statments being factored out into a separate `IsErrorEvictingPodsEvent(...)`-function, so we can have a unit test for this detection methodology specifically, cause i find this to be one of the core features.

Also, I find this particular example easier to read when the conditions are swapped around, so we first check if it's *not* an evictingPods event. So my suggestion would look somewhat like this:

```go
	var newestEvent *corev1.Event

	for _, event := range eventList.Items {
    if ! IsErrorEvictingPodsEvent(event) {
      continue
    }
		if newestEvent == nil {
			newestEvent = &event
			continue
		}
		if newestEvent.LastTimestamp.Time.Before(event.LastTimestamp.Time) {
			newestEvent = &event
		}
	}

```

File: controllers/machine/machine_controller.go
Comment (mrWinston) [inline controllers/machine/machine_controller.go:182]: Why do you pass in the `reqLogger` instead of having the function return an `error`? I think it would make testing easier if we could just check for an error returned instead of parsing the logger output in the unit tests.

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go:174]: Great idea. It would be possible, there's nothing saying you can't cordon a node, drain it, and then uncordon it. That wouldn't replace the node or machine, and then it would work again until it's deleted again.

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go]: Done.

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go]: Good point. Updated for this and the other function.

File: controllers/machine/machine_controller.go
Comment (iamkirkbater) [reply controllers/machine/machine_controller.go:122]: Done.

**Pull request merged**
